### PR TITLE
reformulate bce_with_logits to not use abs

### PR DESCRIPTION
**Pull request by alykhantejani**
This is a fix for the bug reported [here](https://github.com/pytorch/pytorch/pull/1792#issuecomment-317289657) in which `binary_cross_entropy_with_logits` gives the wrong gradient with `input` and `target` are `0`. 

After some investigation, this is because the gradient of `abs` at `0` is `0` (see [here](https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/pointwise.py#L139))

In this PR I have reformulated the numerically stable `binary_cross_entropy_with_logits` to not use `abs` (originally I used what [tensorflow does](https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits), which is use abs).

I guess in general we should think about whether we want the grad of `abs(0)` to be `0`?

### CODE REVIEW

**Comment by fmassa** [comment]
About `grad(abs(0)) = 0`, well, it is not differentiable in `x=0`, and it's sub-derivative can be anything between -1 and 1.
I had a quick look at what TF does, and I have the impression that the gradient at 0 is defined in the same way as in PyTorch, compare [TF abs](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L245) and [PyTorch abs](https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/pointwise.py#L139), and the underlying `sign` function has the same derivative, [TF sign](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L596) and [PyTorch sign](https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/pointwise.py#L363)

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks Aly!
